File: netlogo-gui/src/main/prim/threed/_resizeworld.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.threed

import org.nlogo.api.WorldResizer
import org.nlogo.agent.World3D
import org.nlogo.nvm.{ Command, Context, RuntimePrimitiveException }

class _resizeworld extends Command {

  switches = true



  override def perform(context: Context) {

    val newMinX = argEvalIntValue(context, 0)
    val newMaxX = argEvalIntValue(context, 1)
    val newMinY = argEvalIntValue(context, 2)
    val newMaxY = argEvalIntValue(context, 3)
    val newMinZ = argEvalIntValue(context, 4)
    val newMaxZ = argEvalIntValue(context, 5)

    val oldMinX = workspace.world.minPxcor
    val oldMaxX = workspace.world.maxPxcor
    val oldMinY = workspace.world.minPycor
    val oldMaxY = workspace.world.maxPycor
    val oldMinZ = workspace.world.asInstanceOf[World3D].minPzcor
    val oldMaxZ = workspace.world.asInstanceOf[World3D].maxPzcor

    if (newMinX > 0 || newMaxX < 0 || newMinY > 0 || newMaxY < 0 || newMinZ > 0 || newMaxZ < 0)
      throw new RuntimePrimitiveException(
        context, this,
        "You must include the point (0, 0, 0) in the world.")
    if (oldMinX != newMinX || oldMaxX != newMaxX ||
        oldMinY != newMinY || oldMaxY != newMaxY ||
        oldMinZ != newMinZ || oldMaxZ != newMaxZ) {
      val dimensions =
        new org.nlogo.api.WorldDimensions3D(newMinX, newMaxX,
                                            newMinY, newMaxY,
                                            newMinZ, newMaxZ,
                                            world.patchSize);
      workspace.waitFor(
        new org.nlogo.api.CommandRunnable {
          override def run(): Unit = {
            workspace.setDimensions(dimensions, true, WorldResizer.StopNonObserverJobs)
          }
        })
    }
    context.ip = next
  }
}

File: netlogo-gui/src/main/prim/etc/_resizeworld.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.etc;

import org.nlogo.core.I18N;
import org.nlogo.core.WorldDimensions;
import org.nlogo.api.LogoException;
import org.nlogo.api.WorldResizer;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.RuntimePrimitiveException;

public final strictfp class _resizeworld
    extends Command {
  public _resizeworld() {
    this.switches = true;
  }

  @Override
  public void perform(final Context context) throws LogoException {
    final int newMinX = argEvalIntValue(context, 0);
    final int newMaxX = argEvalIntValue(context, 1);
    final int newMinY = argEvalIntValue(context, 2);
    final int newMaxY = argEvalIntValue(context, 3);

    final int oldMinX = workspace.world().minPxcor();
    final int oldMaxX = workspace.world().maxPxcor();
    final int oldMinY = workspace.world().minPycor();
    final int oldMaxY = workspace.world().maxPycor();

    if (newMinX > 0 || newMaxX < 0 || newMinY > 0 || newMaxY < 0) {
      throw new RuntimePrimitiveException
          (context, this, I18N.errorsJ().get("org.nlogo.prim.etc._resizeworld.worldMustIncludeOrigin"));
    }

    final double patchSize = workspace.world().patchSize();
    final boolean wrappingAllowedInX = workspace.world().wrappingAllowedInX();
    final boolean wrappingAllowedInY = workspace.world().wrappingAllowedInY();

    if (oldMinX != newMinX || oldMaxX != newMaxX ||
        oldMinY != newMinY || oldMaxY != newMaxY) {
      final WorldDimensions dimensions =
        new org.nlogo.core.WorldDimensions(newMinX, newMaxX,
            newMinY, newMaxY, patchSize, wrappingAllowedInX, wrappingAllowedInY);
      workspace.waitFor(
          new org.nlogo.api.CommandRunnable() {
            public void run() {
              workspace.setDimensions(dimensions, true, WorldResizer.stopNonObserverJobs());
            }
          });
    }
    context.ip = next;
  }
}
